### PR TITLE
feat(up): Numeric planning support

### PR DESCRIPTION
**Pull request by arbimo**

### CODE REVIEW

**Review by Shi-Raida — COMMENTED** [review]
edae6a9 is the commit that causes issues on the HDDL problems.

File: solver/src/reasoners/cp/mod.rs [hunk 1]
```diff
@@ -31,56 +32,68 @@ struct LinearSumLeq {
...
+                    .clamp(INT_CST_MIN, INT_CST_MAX),
+            };
+            match domains.present(var) {
+                Some(true) => int_part, // note that if there is no default value, the variable is necessarily present
+                Some(false) => 0,
+                None => 0.min(int_part),
+            }
+        } else {
+            // If None, the var value is considered to be 1
+            elem.factor
```
Comment (arbimo) [inline solver/src/reasoners/cp/mod.rs:55]: What if `elem.lit` is false ? or unset ?

File: solver/src/reasoners/cp/mod.rs [hunk 2]
```diff
@@ -31,56 +32,68 @@ struct LinearSumLeq {
...
+                    .clamp(INT_CST_MIN, INT_CST_MAX),
+            };
+            match domains.present(var) {
+                Some(true) => int_part, // note that if there is no default value, the variable is necessarily present
+                Some(false) => 0,
+                None => 0.max(int_part),
+            }
+        } else {
+            /// If None, the var value is considered to be 1
+            elem.factor
```
Comment (arbimo) [inline solver/src/reasoners/cp/mod.rs:79]: Same comment.

File: solver/src/reasoners/cp/mod.rs [hunk 3]
```diff
@@ -158,26 +178,28 @@ impl Propagator for LinearSumLeq {
     }
 
     fn explain(&self, literal: Lit, domains: &Domains, out_explanation: &mut Explanation) {
-        for e in &self.elements {
-            if e.var != literal.variable() {
-                match e.factor.cmp(&0) {
-                    Ordering::Less => out_explanation.push(Lit::leq(e.var, domains.ub(e.var))),
-                    Ordering::Equal => {}
-                    Ordering::Greater => out_explanation.push(Lit::geq(e.var, domains.lb(e.var))),
+        if domains.entails(self.active) {
```
Comment (arbimo) [inline solver/src/reasoners/cp/mod.rs:181]: Shouldn't `self.active` be in the explanation ?

File: solver/src/reasoners/cp/mod.rs [hunk 4]
```diff
@@ -384,3 +406,87 @@ impl Backtrack for Cp {
...
+/* ========================================================================== */
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    /* ================================= Sum ================================ */
+
+    /// Returns the constraint `x + 2 <= 0`
+    fn simple_constraint() -> (LinearSumLeq, Domains) {
```
Comment (arbimo) [inline solver/src/reasoners/cp/mod.rs:421]: Probably missing some tests with literals other that `TRUE`. 
Checking that you get the expected explanation is also important.

Comment (arbimo) [reply solver/src/reasoners/cp/mod.rs:421]: Also with negative/null factors

**Review by arbimo — COMMENTED** [review]
The above comments are first elements to be considered regarding linear constraints.

I have not yet reviewed in detail the encoding, but at least when applied to scheduling problems they do not seem to work if two activities require the same resource (I am pushing support for scheduling immediately after).

File: solver/src/reasoners/cp/mod.rs (hunk 2 above)
Comment (Shi-Raida) [reply solver/src/reasoners/cp/mod.rs:79]: Same reasoning

File: solver/src/reasoners/cp/mod.rs (hunk 1 above)
Comment (arbimo) [reply solver/src/reasoners/cp/mod.rs:55]: I think indeed it is wrong for both. 
I only looked at it quickly, but I believe that 

 - At line 48, you should not branch on the presence literal but on the `elem.lit` literal.
 - you should the same thing in the else branch (before this patch it was done coherently in both cases).

Comment (arbimo) [reply solver/src/reasoners/cp/mod.rs:55]: The question really is: what is the lowest possible value that this term (`lit * factor *var`) may have?

Comment (Shi-Raida) [reply solver/src/reasoners/cp/mod.rs]: Fixed in 4bf4851

File: solver/src/reasoners/cp/mod.rs (hunk 2 above)
Comment (Shi-Raida) [reply solver/src/reasoners/cp/mod.rs]: Fixed in 4bf4851

File: solver/src/reasoners/cp/mod.rs (hunk 4 above)
Comment (Shi-Raida) [reply solver/src/reasoners/cp/mod.rs]: The code coverage has been increased in the last commits

**Comment by Shi-Raida** [comment]
I needed to create the `VarRef::ONE` variable in order to fix all issues.

In particular, it was not possible to return an `InvalidUpdate` in the `set_ub()` function.
The `InvalidUpdate` needs a literal which is formed with the `SignedVar` and the upper bound.
However, in the case of a constant term, there was no variable and so no `SignedVar`.

With this new variable always equals to 1, all terms of a linear sum can be treated the same way.

**Pull request merged**